ring/ring.go

```diff
@@ -705,11 +707,6 @@ func (r *Ring) ShuffleShard(identifier string, size int) ReadRing {
 // This function supports caching, but the cache will only be effective if successive calls for the
 // same identifier are with the same lookbackPeriod and increasing values of now.
 func (r *Ring) ShuffleShardWithLookback(identifier string, size int, lookbackPeriod time.Duration, now time.Time) ReadRing {
-	// Nothing to do if the shard size is not smaller then the actual ring.
-	if size <= 0 || r.InstancesCount() <= size {
```

Comment:
It's pity we lose this optimization. What can we do here?

Reply:
But why was removed in the first place? If shuffle sharding is disabled, we should always returns the full ring, isn't it?

Reply:
I think this optimization can be restored as well. Thanks for raising this.

Reply:
`ShuffleShardWithLookback` now filters out read-only instances with `ReadOnlyUpdatedTimestamp` outside of lookback period. If we return entire ring, read-only instances will remain in the ring.

Filtering out read-only instances outside of lookback window is an optimization and not strictly required. Returning entire ring is a valid choice.

I think we should make code consistent:
* either remove this shortcut, and keep filtering of read-only instances with state change outside of lookback window
* or remove the filtering (which I added yesterday)

I don't have strong opinion, and I'm fine if we remove the filtering, and always return read-only instances (but still extend the shard!) when lookback period is > 0.

Reply:
I'll go ahead and remove the filtering mentioned in my comment.

Reply:
Done in https://github.com/grafana/dskit/pull/554/commits/a29a1ea7035791d19113d2638b5fb4c55090fcc4.

Reply:
> @pracucci voiced concern about NOT filtering out such instances:
> 
> > Isn't there the risk of failing quorum, due to race conditions with shutdown? Like shutdown a read-only instance that was just returned by `ShuffleShardWithLookback()`
> 
> And also pointed out that we _do_ filter out read-only partitions on shuffle shard with lookback.

Let's address this in the next PR.